Cleanup part file after upload exception

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/17149
- [x] write unit tests
- [x] TEST: upload of small file + permission error on fopen
- [x] TEST: upload of small file + exception during final rename
- [x] TEST: upload of small file + exception during stream copy
- [x] TEST: upload of big file + exception during final rename
- [x] TEST: upload of big file + exception during chunk assembly

@icewind1991 @MorrisJobke 
(I haven't tested this yet)

Commit:
Cleanup part file after upload exception

Added unit tests for checking for stray part files.
Convert exception to sabre exception in upload put method.

Also added unit test for exception mapping, which also indirectly tests
that the part file is being deleted on exception.

This applies to both chunking and non-chunking mode.

Added some unit tests for chunk upload.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9359e2c3-b1f8-4b0a-bf42-05082553e41c) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/95ef82e4-9deb-4ae4-bd83-90e55d5234a8) was created.

PVince81 commented:
Added some unit tests for small files.
But I'm not happy in general because we have a lot of exceptions.
We should refactor this to catch any exception and convert it with a new method `convertToSabreException()` before rethrowing.

Also there are no unit tests for big file chunking, so testing exceptions and part files there will be difficult.

PVince81 commented:
But one cool thing is how I'm taking PHPUnit mocks to the next level: https://github.com/owncloud/core/pull/17189/files#diff-06f023f104a1314e2e8fdfebf767e963R51 :wink:

PVince81 commented:
Raised https://github.com/owncloud/core/issues/17220 to refactor exception handling.

DeepDiver1975 commented on lib/private/connector/sabre/file.php:
```diff
@@ -147,25 +147,36 @@ public function put($data) {
 			}
 
 		} catch (NotPermittedException $e) {
+			$partStorage->unlink($internalPartPath);
```
hmm - this kind of cries already for the 'global' exception handling/conversion function .... :speak_no_evil:

LukasReschke replied:
I should patent my Base.php idea:
```php
<?php
try {
// All of base.php
} catch (\Exception $e) {
echo "All fine. Nothing to worry about.";
}
```

LukasReschke replied:
:speak_no_evil: :hear_no_evil: :see_no_evil:

PVince81 replied:
Yes it does... but I'm afraid to refactor at this point as we don't have enough test coverage... (and also not enough time)

PVince81 replied:
Maybe another idea would be to use a shutdown handler to clear part files. This means they need to be registered somewhere to be cleant up later... might be more complicated.

PVince81 commented:
Also to note: I'm writing a bit of unit tests for chunking, because without even some base tests it will be impossible to test chunk clearing...

PVince81 commented:
Looks like refactoring will be inevitable: the chunking code path doesn't translate half of the exceptions...

I'll limit the refactoring to the specific try/catch blocks that clean up chunks to limit side-effects.
Ideally in the future all try/catch blocks should have that logic.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/945356ae-e6c1-4e18-9add-8d36443b791f) was created.

PVince81 commented:
Ok, I added unit test for the file chunking cases and also refactored exception conversion into a method.

Please have a look and tell me it's a bit saner now...
Will still need some manual testing.

MorrisJobke commented on lib/private/connector/sabre/file.php:
```diff
@@ -426,4 +411,47 @@ private function needsPartFile($storage) {
+			throw new ServiceUnavailable("Failed to write file contents: " . $e->getMessage());
+		}
+
+		throw new \Sabre\DAV\Exception($e->getMessage());
```
can you please add the previous exception as third parameter. Makes traces more easy.

PVince81 replied:
Ok, I'll try and modify all the exception classes to accept a $reason argument.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7d8925d9-36c3-494a-89a6-531571346784) was created.

PVince81 commented:
Done: I added the previous exception to the constructors

PVince81 commented:
- [ ] BUG: part files not deleted for chunk upload with locking enabled

MorrisJobke commented:
```
00:36:24.357 There was 1 failure:
00:36:24.357 
00:36:24.357 1) OC\UpdaterTest::testCheckWithEmptyInvalidXmlResponse
00:36:24.357 Expectation failed for method name is equal to <string:setAppValue> when invoked at sequence index 1
00:36:24.357 Parameter 2 for invocation OCP\IConfig::setAppValue('core', 'lastupdatedat', 1435583499) does not match expected value.
00:36:24.357 Failed asserting that 1435583499 matches expected 1435583498.
00:36:24.357 
00:36:24.357 /var/jenkins/workspace/pull-request-analyser-ng-simple@3/label/SLAVE/lib/private/updater.php:135
00:36:24.357 /var/jenkins/workspace/pull-request-analyser-ng-simple@3/label/SLAVE/tests/lib/updater.php:345
```

PVince81 commented:
Ok, the locking issue seems to be that whenever an exception occurs during the final rename in chunking mode, the part file is not getting unlocked yet. So it's not possible to delete it.

Needs further research.
What wonders me is why the unit tests don't catch this case.

Commit:
Unlock first path on rename if second path is locked

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6e0d7d16-1b34-4f54-a813-50d3a9040e32) was created.

PVince81 commented:
Squashed and fixed the unit tests to properly catch the locking exception.

I also added a fix for $view->rename to properly unlock the source path (here the part file), so we can delete it after failure: https://github.com/owncloud/core/commit/621bebdd4affb15ecf1637fb545afed3730b009e

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9e4915e0-5762-4be8-8fc5-346cc864e845) was created.

PVince81 commented:
Rebased onto master to avoid the class name issue.

This is ready for review @MorrisJobke @nickvergessen @icewind1991 @DeepDiver1975 @schiesbn

icewind1991 commented:
Looks good :+1:

nickvergessen commented on lib/private/connector/sabre/file.php:
```diff
@@ -207,9 +192,9 @@ public function put($data) {
 						$partStorage->unlink($internalPartPath);
```
Since we throw an exception right below, we will end up in the catch where we try to re-unlink the file?

PVince81 replied:
True, missed it. Thanks.

nickvergessen commented on lib/private/connector/sabre/file.php:
```diff
@@ -399,10 +387,11 @@ private function createFileChunked($data) {
-			} catch (LockedException $e) {
-				throw new FileLocked($e->getMessage(), $e->getCode(), $e);
+			} catch (\Exception $e) {
+				if ($partFile) {
```
=== please

Which of `''`, `null`, `'0'` do you want?

MorrisJobke replied:
@PVince81 Please address this

PVince81 replied:
I left it out because it's always ".part", even with "0" it would be "0.part".

Anyway, I'll need to touch this PR for the other comments so I'll fix this as well.

nickvergessen commented on lib/private/files/view.php:
```diff
@@ -631,7 +631,12 @@ public function rename($path1, $path2) {
+			try {
+				$this->lockFile($path2, ILockingProvider::LOCK_SHARED, true);
+			} catch (LockedException $e) {
+				$this->unlockFile($path1, ILockingProvider::LOCK_SHARED);
```
Ah, the stray lock :)

MorrisJobke commented:
Still works here. There are some additional things I need to debug, but nothing that blocks this and can be done in a followup PR 

:+1:

MorrisJobke commented:
-.- Now github showed my the comments of @nickvergessen

PVince81 commented:
Thanks for merging :smile: 

I'll send a follow up PR for the small stuff.

MorrisJobke commented:
> I'll send a follow up PR for the small stuff.

@PVince81 Ping

PVince81 commented:
It's on my internal todo list. Was busy with more important locking stuff yesterday.

PVince81 commented:
Follow up PR here https://github.com/owncloud/core/pull/17268
Please review.
